src/core/tools/common/thrift_message_parser.cpp

```diff
@@ -128,33 +116,38 @@ namespace dsn
         dassert(msg->header->server.error_name[0], "error name should be set");
         dassert(!msg->buffers.empty(), "buffers can not be empty");
 
-        // response format:
-        //     <total_len(int32)> <thrift_string> <body_data(bytes)>
-        //    |-----------response header--------|
+        // write thrift response header and thrift message begin
         binary_writer header_writer;
-        binary_writer_transport trans(header_writer);
-        boost::shared_ptr<binary_writer_transport> trans_ptr(&trans, [](binary_writer_transport*) {});
-        ::apache::thrift::protocol::TBinaryProtocol proto(trans_ptr);
-
-        //Total length, but we don't know the length, so firstly we put a placeholder
-        proto.writeI32(0);
+        binary_writer_transport header_trans(header_writer);
```

Comment:
can we move these logic into prepare_on_send as much as possible to reduce the lock conflict?